src/components/button/button.js

```diff
@@ -0,0 +1,70 @@
+// javascript 'shim' to trigger the click event of element(s)
+// when the space key is pressed.
+//
+// Created since some Assistive Technologies (for example some Screenreaders)
+// Will tell a user to press space on a 'button', so this functionality needs to be shimmed
+// See https://github.com/alphagov/govuk_elements/pull/272#issuecomment-233028270
+//
+// Usage instructions:
+// the 'shim' will be automatically initialised
+;(function (global) {
+  'use strict'
+
+  var GOVUK_FRONTEND = global.GOVUK_FRONTEND || {}
+
+  GOVUK_FRONTEND.buttons = {
+
+    // Add event construct for modern browsers or IE
+    // which fires the callback with a pre-converted target reference
+    addEvent: function (node, type, callback) {
+      if (node.addEventListener) {
+        node.addEventListener(type, function (e) {
+          callback(e, e.target)
+        }, false)
+      } else if (node.attachEvent) {
+        node.attachEvent('on' + type, function (e) {
+          callback(e, e.srcElement)
+        })
+      }
+    },
+
+    // Cross-browser character code / key pressed
+    charCode: function (e) {
+      return (typeof e.which === 'number') ? e.which : e.keyCode
+    },
+
+    // Cross-browser preventing default action
+    preventDefault: function (e) {
+      if (e.preventDefault) {
+        e.preventDefault()
+      } else {
+        e.returnValue = false
+      }
+    },
+
+    eventHandler: function (event) {
+      // if the keyCode (which) is 32 it's a space, let's simulate a click.
+      if (GOVUK_FRONTEND.buttons.charCode(event) === 32) {
+        GOVUK_FRONTEND.buttons.preventDefault(event)
+        // trigger the target's click event
+        event.target.click()
+      }
+    },
+
+    init: function () {
+      var buttons = document.querySelectorAll('[role="button"]')
```

Comment:
This will only work for buttons that are present on document ready, not buttons that are loaded in afterwards (e.g. via an AJAX request). This would be a 'regression' from what we've got in Frontend Toolkit, where we use a delegated event handler.

This might be OK, but we might want to consider:

- documenting this limitation
- allow `init` to be run more than once (i.e. ensure we only ever bind to the buttons that we haven't already handled)?

Reply:
I believe jQuery uses global document event handling to work around this.

e.g.
```js
document.addEventListener('click', event => {
  if (event.target === rolebutton) {
  }
})
```

Reply:
@36degrees yep, it's something that I'm looking into.
I was thinking that we can either reinitialise on `DOMContentLoaded` on `document`
`addEvent(document, 'DOMContentLoaded', init)` or watch for the `click` event at the document level as @nickcolley described.